### ISSUE TITLE
New Ensemble Forecasting Methods

### ISSUE THREAD

**Issue by magittan** [alan-turing-institute/sktime#271]
**Is your feature request related to a problem? Please describe.**
Add new ensemble forecasting methods that would also involve forecasting in an online setting.

**Describe the solution you'd like**
Would like to implement framework to add new ensemble algorithms that focus on choosing a probability distribution over the forecasters, would involve adding ensemble algorithms that would be composed with the existing EnsembleForecaster(BaseForecaster) and not affect current implementations.

![Expert UML](https://user-images.githubusercontent.com/14024202/83341789-b2287e80-a2b5-11ea-89b2-d60ffa29c201.png)

Some of these algorithm implementations would be a Non-Negative Least Squares Fit, Hedge-based algorithms, and another type of ensemble algorithm based on [Discrepancy](http://proceedings.mlr.press/v49/kuznetsov16.html)

**Additional context**
Forecasting in an online setting may not be applicable to non-stationary time series and the algorithms used would not give optimal regret bounds in these cases, however, it could be useful to have and some methods can handle non-stationary time series. 

(I have not considered alternatives to this but would be welcome to any other suggestions)

**Comment by mloning** [comment]
Thanks again for getting in touch and opening the issue here! 

A few comments: 
* We have an interface for updating forecasters: `update(y_new, X_new=None, ...)`, perhaps we can re-use that? 
* Could you add some references for the other ensemble algorithms? 
* What's the output of `online_predict()` and how would that be different from calling `update()` and `predict()` sequentially?

**Comment by magittan** [comment]
- For the interface for updating forecasters, yes we can totally re-use that
- For a Hedge based algorithm, I was hoping to implement https://cseweb.ucsd.edu/~yfreund/papers/nhedge.pdf and for NNLS scipy offers an implementation https://docs.scipy.org/doc/scipy-0.14.0/reference/generated/scipy.optimize.nnls.html. There are other best expert prediction algorithms that are better for non-stationary settings that I'm trying to read up on https://users.soe.ucsc.edu/~manfred/pubs/J39.pdf
-  The output of online prediction would be the prediction for some input X but the catch is that after each prediction, the update is performed. It would definitely be possible to do update() and predict() sequentially and the function itself is really just a faster way of implementing that so I think leaving it out would be good.

**Comment by mloning** [comment]
We also have an `update_predict_single` method if there are more efficient implementations for updating and predicting. There's also an `update_predict` function for model evaluation purposes.

**Comment by magittan** [comment]
Sorry, I couldn't find this method within the BaseForecaster, BaseUpdateableForecaster, or EnsembleForecaster classes. Could the functionality of update_predict be extended to the EnsembleForecaster class?

**Comment by magittan** [comment]
Would unit tests be required for each of the ensemble algorithms? I have some examples in the forms of notebooks to show qualitative behavior but nothing that can be run on a script for exact values?

**Comment by mloning** [comment]
Ah sorry, most of the forecasting functionality is still on the `dev` branch, we'll push it to master soon. Unit tests would be needed, but it's easy to write them up if you already have some examples.

**Comment by magittan** [comment]
I see, in the meantime, I can work on adding different algorithms and further tuning some examples?

**Comment by mloning** [comment]
Yes sounds good!

**Comment by magittan** [comment]
Hi, I was trying to implement an online training method by rewriting the update_predict with a cv that will split my data into parts where I update the forecaster with the current step and predict the next step and I was wondering how I would implement such a cv. I was thinking that something like:

Input:
y_train = [1,2,3,4,5,6,7,8,9]

Output:
```python
for new_window, _ in cv.split(y_train):
   print(new_window)
   print(_)
```

Expected Partitions:
[1]
[2]
[1,2]
[3]
[1,2,3]
[4]
....

I'm mostly looking at the code for `_predict_moving_cutoff` from _sktime.py, I want to understand how `_update_predict_single`

But based on the update step, I wasn't sure how the update saves the new window. I was thinking maybe a cv with the output:

Expected Partitions:
[1]
[2]
[2]
[3]
[3]
[4]

Would work if the update step saved the previous step.

**Comment by mloning** [comment]
Hi @magittan, we've got a sliding window splitter: 

```python
from sktime.forecasting.model_selection import SlidingWindowSplitter
from sktime.datasets import load_airline

y = load_airline().iloc[:10]
cv = SlidingWindowSplitter(start_with_window=True, window_length=5, fh=[1, 2])
for train_window, test_window in cv.split(y):
    print(train_window, test_window)

> [0 1 2 3 4] [5 6]
> [1 2 3 4 5] [6 7]
> [2 3 4 5 6] [7 8]
> [3 4 5 6 7] [8 9]
```
Does that help?

**Comment by magittan** [comment]
I guess my questions is more about how the forecaster's update function works. I see that it updates the `oh` parameter with the total amount of data seen so far, but does it also fit on that data?

**Comment by mloning** [comment]
that's controlled by the `update_params` kwargs. Many algorithms allow to updated fitted parameters incrementally from new data, without having to refit the whole model. Updating is often also more efficient than refitting. For example, an ARIMA model can be updated using Kalman smoothing. The idea is to implement these methods inside the update functionality. So, if you set `update_params=True`, fitted parameters and as well as the observation horizon (`oh`) would be updated, otherwise only the observation horizon would be updated.

**Comment by magittan** [comment]
I see, this model definitely makes sense. I guess I was just unsure of how (or if even) to allow the forecasters to refit when data comes in. I think I'll first consider the simpler case where no refit is needed for the forecaster and rather the ensemble will fit itself during a validation stretch of data.

**Comment by mloning** [comment]
Closed by #333
